include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -2518,6 +2518,32 @@ namespace internal
   }
 
 
+
+  /**
+   * TODO
+   */
+  template <int dim, typename Number, typename Number2>
+  inline void
+  integrate_tensor_product_value_and_gradient(
```

Comment:
@kronbichler This is still missing. Any recommendations how to tackle this function?

Reply:
It should be the transpose of the other function. So you first evaluate the polynomial basis in all points as above, and then switch the left hand side and right hand sides of the assignments. Here is the example of what I believe is needed in 2D (without renumbering and without cleanup code in case the derivative or value is empty; 1D should 3D be analogous):
```c++
  for (unsigned int i1 = 0, i = 0; i1 < (dim > 1 ? n_shapes : 1); ++i1)
    {
      Number test_value_y = value * shapes[2 * n_shapes + 2 * i1] +
                            gradient[1] * shapes[2 * n_shapes + 2 * i1 + 1];
      Number test_grad_y = gradient[1] * shapes[2 * n_shapes + 2 * i1];
      for (unsigned int i0 = 0; i0 < n_shapes; ++i0, ++i)
        values[i] = shapes[2 * i0] * test_value_y + shapes[2 * i0 + 1] * test_grad_y;
  }
```